library/os_container_infra.py

```diff
@@ -0,0 +1,136 @@
+#!/usr/bin/env python
+#
+# This script allows a user to create, scale (update)  or get inventory of a
+# Magnum cluster. It is essential that a cluster template already exists before
+# cluster creation or scaling is requested.
+#
+# To view warnings emitted by this script, export the following
+# following variable before running ansible:
+#
+#   export ANSIBLE_LOG_PATH=ansible.log
+#
+# To filter these warnings:
+#
+#   tail -f ansible.log | grep WARNING
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.utils.display import Display
+from magnumclient.client import Client
+from magnumclient.exceptions import *
+import openstack
+import time
+
+display = Display()
+
+def get_os_cloud(cloud_name, module):
+    os_cloud = openstack.connect(cloud=cloud_name)
+    os_cloud.authorize()
+    return os_cloud
+
+def get_update_or_create_container_infra(magnum_client, module, **kwargs):
```

Comment:
**No need to change**, but the swiss army knife function approach can be a little hard to follow, and means that every poll iteration has the possibility of creating or updating the cluster, when it should just be querying. Consider something like this, with the logic factored out to the top level:

* Does the cluster exist?
* If we want it to exist:
  * If it does not exist, create and wait for completion or failure
  * If it does exist, update and wait for completion or failure
* If we do not want it to exist:
  * If it does exist, delete it and wait for deletion

Reply:
I agree, it is the first iteration and this function is a bit of a mess!